src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1428 @@
+                BoundSlicePattern => default,
+                BoundITuplePattern => default,
+                BoundNegatedPattern => default,
+                BoundTypePattern => default,
+                BoundDeclarationPattern => default,
+                BoundConstantPattern => default,
+                BoundDiscardPattern => default,
+                BoundRelationalPattern => default,
+                null => default,
+                _ => throw ExceptionUtilities.UnexpectedValue(pattern)
+            };
+
+            static SetsOfOrCases rewriteBinary(BoundBinaryPattern binaryPattern, Conversions conversions)
+            {
+                SetsOfOrCases result = default;
+
+                if (binaryPattern.Disjunction)
+                {
+                    var patterns = ArrayBuilder<BoundPattern>.GetInstance();
+                    addPatternsFromOrTree(binaryPattern, patterns);
+
+                    // In `A1 or ... or An`, we produce an expansion set: `case A1`, ..., `case An`
+                    SetOfOrCases resultOrSet1 = result.StartNewOrCases();
+                    var inputType = binaryPattern.InputType;
+                    foreach (var pattern in patterns)
+                    {
+                        resultOrSet1.Add(pattern, pattern.Syntax);
+                    }
+
+                    // In `A1 or ... or An or B or ...` where `B` has an expansion set: `case B1`, ..., `case Bn`
+                    // we produce an expansion set: `case A1:`, ... `case An:`, `case B1:`, ... `case Bn:`
+                    for (int i = 0; i < patterns.Count; i++)
+                    {
+                        BoundPattern pattern = patterns[i];
+                        using SetsOfOrCases setOfOrCases = RewriteToSetsOfOrCases(pattern, conversions);
+                        if (!setOfOrCases.IsDefault)
+                        {
+                            foreach (SetOfOrCases orSet in setOfOrCases.Set)
+                            {
+                                SetOfOrCases resultOrSet = result.StartNewOrCases();
+
+                                for (int j = 0; j < i; j++)
+                                {
+                                    resultOrSet.Add(patterns[j], syntax: null); // already checked reachability in the previous set
+                                }
+
+                                foreach ((BoundPattern resultPattern, SyntaxNode? syntax) in orSet.Cases)
+                                {
+                                    resultOrSet.Add(resultPattern, syntax);
+                                }
+                            }
+                        }
+                    }
+
+                    patterns.Free();
+                }
+                else
+                {
+                    // In `A and B`, when found multiple `or` patterns in `A` to be expanded, we drop the `and B`
+                    result = RewriteToSetsOfOrCases(binaryPattern.Left, conversions);
```

Comment:
>result = RewriteToSetsOfOrCases(binaryPattern.Left, conversions); [](http://example.com/codeflow?start=20&length=65)

It looks like here is another place where we can hit a deep recursion on the left.